text/0000-overload-logic-operators.md

```diff
@@ -0,0 +1,228 @@
+[guide-level-explanation]: #guide-level-explanation
+
+This proposal starts with an enum definition and trait definitions for each of the operators:
+
+```rust
+enum ShortCircuit<S, L> {
+    Short(S),
+    Long(L),
+}
+
+trait LogicalOr<Rhs = Self>: Sized {
+    type Output;
+    type Intermediate;
+
+    /// Decide whether the *logical or* should short-circuit
+    /// or not based on its left-hand side argument. If so,
+    /// return its final result, otherwise return the value
+    /// that will get passed to `logical_or()` (normally this
+    /// means returning self back, but you can change the value).
+    fn short_circuit_or(self) -> ShortCircuit<Self::Output, Intermediate>;
+
+    /// Complete the *logical or* in case it did not short-circuit.
+    /// Normally this would just return `rhs`.
+    fn logical_or(lhs: Intermediate, rhs: Rhs) -> Self::Output;
+}
+
+trait LogicalAnd<Rhs = Self>: Sized {
+    type Output;
+    type Intermediate;
+
+    /// Decide whether the *logical and* should short-circuit
+    /// or not based on its left-hand side argument. If so,
+    /// return its final result, otherwise return the value
+    /// that will get passed to `logical_and()` (normally this
+    /// means returning self back, but you can change the value).
+    fn short_circuit_and(self) -> ShortCircuit<Self::Output, Intermediate>;
+
+    /// Complete the *logical and* in case it did not short-circuit.
+    fn logical_and(lhs: Intermediate, rhs: Rhs) -> Self::Output;
+}
+```
+
+With a matching desugaring:
+
+```rust
+<expr_a::<T>> || <expr_b::<T>>
+
+==>
+
+match expr_a.short_circuit_or() {
+    ShortCircuit::Short(res) => res,
+    ShortCircuit::Long(lhs) => logical_or(lhs, expr_b)
+}
+```
+
+and
+
+```rust
+<expr_a::<T>> && <expr_b::<T>>
+
+==>
+
+match expr_a.short_circuit_and() {
+    ShortCircuit::Short(res) => res,
+    ShortCircuit::Long(lhs) => logical_and(lhs, expr_b)
+}
+```
+
+From taking into consideration the current functions, and previous discussion on [internals](https://internals.rust-lang.org/t/pre-rfc-overload-short-curcuits/10460) it seems that the following makes the most sense in terms of outcomes.
+
+#### For `Option<T>`:
+
+```rust
+fn foo() -> Option<i32> {
+    Some(3)
+}
+
+fn main() {
+    Some(4) || Some(5); // == Some(4)
+    None || Some(5);    // == Some(5)
+    Some(4) || foo();   // == Some(4) (foo is *not* called)
+    None || foo();      // == Some(3) (foo is called)
+    None || 3;          // == 3
+    Some(2) || 1;       // == 2
+    Some(1) || panic!() // == Some(1)               + These two are side effects from !
+    None || return      // returns from function    + and are the same to how boolean || works
+    Some(2) && Some(3)  // Some(3)
+    None && Some(1)     // None
+    Some(3) && None     // None
+
+    Some(2) || Some("hello") // Error: LogicalOr<Option<&str>> not implemented for Option<i32>
+    Some(2) || 2 || 3        // Error: LogicalOr<i32> is not implemented for i32
```

Comment:
```suggestion
    // or
    Some(4) || Some(5); // == Some(4)
    None || Some(5);    // == Some(5)
    // or_else
    Some(4) || foo();   // == Some(4) (foo is *not* called)
    None || foo();      // == Some(3) (foo is called)
    // unwrap_or
    None || 3;          // == 3
    Some(2) || 1;       // == 2
    // weird
    Some(1) || panic!() // == Some(1)               + These two are side effects from !
    None || return      // returns from function    + and are the same to how boolean || works
    // and
    Some(2) && Some(3)  // Some(3)
    None && Some(1)     // None
    Some(3) && None     // None
    // and_then (without taking parameter)
    Some(4) && foo();   // None (foo is called)
    None && foo();      // None (foo is **not** called)

    Some(2) || Some("hello") // Error: LogicalOr<Option<&str>> not implemented for Option<i32>
    Some(2) || 2 || 3        // Error: LogicalOr<i32> is not implemented for i32
```

```
    Some(1) || panic!() // == Some(1)               + These two are side effects from !
    None || return      // returns from function    + and are the same to how boolean || works
```
Looks like magic, what are these? How about `Some(3) || return`?

How about `Option` to `Result` type?
```
Some(4) && Ok(5)
None && Ok(5)
Some(4) && Err("no")
None && Err("no")
Some(4) || Ok(5)
None || Ok(5)
Some(4) || Err("no")
None || Err("no")
```

Reply:
> `Some(3) || return`

`Some(3) || return == Some(3)` similar to how `Some(1) || panic!() == Some(1)`

> How about `Option` to `Result` type? And vice versa?

There are already functions to do that conversion, so it doesn't seem necessary.